pkg/server/server.go

```diff
@@ -145,23 +154,61 @@ func (c *Server) Run() {
 	errLog := logrus.WithField("http", "error").Writer()
 	defer errLog.Close()
 
+	handler := c.getHandler()
+
 	logrus.Infof("listening on %s", listenURL)
-	logrus.Infof("reconfigure your apiserver with `--authentication-token-webhook-config-file=%s` to enable (assuming default hostPath mounts)", c.GenerateKubeconfigPath)
-	httpServer := http.Server{
+	logrus.Infof("configure your apiserver with `--authentication-token-webhook-config-file=%s` to enable (assuming default hostPath mounts)", c.GenerateKubeconfigPath)
+	httpServer := &http.Server{
 		ErrorLog: log.New(errLog, "", 0),
-		Handler:  c.getHandler(),
+		Handler:  handler,
+	}
+
+	go func() {
+		if err := httpServer.Serve(listener); err != http.ErrServerClosed {
+			logrus.WithError(err).Fatal("http server exited")
+		}
+	}()
+
+	ctrl, err := controller.New(c.Config, handler.mappings)
+	if err != nil {
+		logrus.WithError(err).Fatal("controller failed to initialize")
+	}
+	ctrl.StartWatch()
+
+	handleShutdown(httpServer, 3*time.Second)
+}
+
+func handleShutdown(hs *http.Server, timeout time.Duration) {
+	stop := make(chan os.Signal, 1)
+
+	signal.Notify(stop, os.Interrupt, syscall.SIGTERM)
+
+	<-stop
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	logrus.Infof("shutdown with timeout: %s\n", timeout)
+
+	if err := hs.Shutdown(ctx); err != nil {
+		logrus.WithError(err).Fatal("shutdown error")
+	} else {
+		logrus.Infof("server stopped")
 	}
-	logrus.WithError(httpServer.Serve(listener)).Fatal("HTTP server exited")
 }
 
 func (c *Server) getHandler() *handler {
 	h := &handler{
+		config:           c.Config,
 		lowercaseRoleMap: make(map[string]config.RoleMapping),
 		lowercaseUserMap: make(map[string]config.UserMapping),
 		accountMap:       make(map[string]bool),
 		verifier:         token.NewVerifier(c.ClusterID),
 		metrics:          createMetrics(),
 		ec2Provider:      newEC2Provider(c.ServerEC2DescribeInstancesRoleARN),
+		mappings: &mappings.Mappings{
```

Comment:
Create a constructor for Mappings `mapping.New` that initializes Identities for you.